### PR TITLE
rf cookbook

### PR DESCRIPTION
**Pull request by Saurabh7**

### CODE REVIEW

**Comment by Saurabh7** [comment]
Needs a fix for python3 https://github.com/shogun-toolbox/shogun/pull/3302

File: doc/cookbook/source/examples/classifier/random_forest.rst [hunk 1]
```diff
@@ -0,0 +1,38 @@
+============================
+Random Forest classification
+============================
+
+Random Forest is an ensemble learning method which implements multiple decision trees while training and performs prediction by using a combination rule on the outputs of individual decision trees.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/random_forest.rst]: Do "...during training. It predicts by using..."

File: doc/cookbook/source/examples/classifier/random_forest.rst [hunk 2]
```diff
@@ -0,0 +1,38 @@
+============================
+Random Forest classification
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/random_forest.rst]: remove classification here, no need

File: doc/cookbook/source/examples/classifier/random_forest.rst [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+-------
+Example
+-------
+
+CDenseFeatures (here 64 bit floats aka RealFeatures) and :sgclass:`CMulticlassLabels` are created from training and test data file
+
+.. sgexample:: random_forest.sg:create_features
+
+A combination rule of :sgclass:`CMajorityVote` is created to be used as a combination rule. This and other combination rules are derived form the :sgclass:`CCombinationRule` .
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/random_forest.rst]: Can you maybe phrase all this in the way that there is a default configuration, and then *optionally* one can put in these decision rules. See what I mean?

File: doc/cookbook/source/examples/classifier/random_forest.rst [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+Next an instance of :sgclass:`CRandomForest` is created. The parameters provided are the number of attributes to be chosen randomly to select from and the number of trees.
+
+.. sgexample:: random_forest.sg:create_instance
+
+Then we run the train random forest and apply it to test data, which here gives :sgclass:`CMulticlassLabels`.
+
+.. sgexample:: random_forest.sg:train_and_apply
+
+We can evaluate test performance via e.g. :sgclass:`CMulticlassAccuracy` as well as get the Out Of Bag Error.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/random_forest.rst]: "out of bag error" ... use ""

File: doc/cookbook/source/examples/classifier/random_forest.rst [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+.. sgexample:: random_forest.sg:evaluate_accuracy
+
+----------
+References
+----------
+:wiki:`Random_forest`
+:wiki:`Out-of-bag_error`
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/random_forest.rst]: need a newline

File: examples/meta/src/classifier/random_forest.sg [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+rand_forest.set_labels(labels_train)
+#![create_instance]
+
+#![train_and_apply]
+rand_forest.train(feats_train)
+MulticlassLabels labels_predict = rand_forest.apply_multiclass(features_test)
+#![train_and_apply]
+
+#![evaluate_accuracy]
+MulticlassAccuracy eval()
```
Comment (karlnapf) [inline examples/meta/src/classifier/random_forest.sg]: pls dont call this eval, it is a python keyword

File: doc/cookbook/source/examples/classifier/random_forest.rst [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+.. bibliography:: ../../references.bib
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/random_forest.rst]: can we have some references, e.g. on the first page here: http://civil.colorado.edu/~balajir/CVEN6833/lectures/cluster_lecture-2

**Comment by Saurabh7** [comment]
updated!
https://github.com/shogun-toolbox/shogun-data/pull/106

**Comment by karlnapf** [comment]
Merged data, you can update it in here

File: doc/cookbook/source/examples/classifier/random_forest.rst [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+-------
+Example
+-------
+
+CDenseFeatures (here 64 bit floats aka RealFeatures) and :sgclass:`CMulticlassLabels` are created from training and test data file
+
+.. sgexample:: random_forest.sg:create_features
+
+Combination rules to be used for prediction are derived form the :sgclass:`CCombinationRule` class. Here we create a :sgclass:`CMajorityVote` class to be used as a combination rule.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/random_forest.rst]: BTW. Does RF work if no combination rule is supplied?
It should by default, and then you should say here that changing it is optional

File: examples/meta/src/classifier/random_forest.sg [hunk 9]
```diff
@@ -0,0 +1,36 @@
+CSVFile f_feats_train("../../data/classifier_4class_2d_linear_features_train.dat")
+CSVFile f_feats_test("../../data/classifier_4class_2d_linear_features_test.dat")
```
Comment (karlnapf) [inline examples/meta/src/classifier/random_forest.sg]: The test will fail if you dont fix seed I think

Comment (karlnapf) [inline examples/meta/src/classifier/random_forest.sg]: So you might need to resend the integration test data with a fixed seed

File: doc/cookbook/source/examples/classifier/random_forest.rst [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+Random Forest is an ensemble learning method which implements multiple decision trees during training. It predicts by using a combination rule on the outputs of individual decision trees.
+
+See :cite:`Breiman2001` for a detailed introduction.
+
+-------
+Example
+-------
+
+CDenseFeatures (here 64 bit floats aka RealFeatures) and :sgclass:`CMulticlassLabels` are created from training and test data file
```
Comment (vigsterkr) [inline doc/cookbook/source/examples/classifier/random_forest.rst]: please make it sure that we can handle CombinedFeatures where we can use both Dense and String features... remember it's a tree!

Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/random_forest.rst]: AFAIK, the current implementation cannot do that. 
The Random forest interface can, but the CartTree inside assumes float64_t.
Should definitely be changed, see email I wrote. 
We can (and should) update this cookbook once Shogun supports these things

File: doc/cookbook/source/examples/classifier/random_forest.rst (hunk 8 above)
Comment (Saurabh7) [inline doc/cookbook/source/examples/classifier/random_forest.rst]: No it doesn't right now.

**Comment by Saurabh7** [comment]
I have fixed seed. Sry abt that.
https://github.com/shogun-toolbox/shogun-data/pull/111

File: doc/cookbook/source/examples/multiclass_classifier/random_forest.rst [hunk 11]
```diff
@@ -0,0 +1,41 @@
+=============
+Random Forest
+=============
+
+Random Forest is an ensemble learning method which implements multiple decision trees during training. It predicts by using a combination rule on the outputs of individual decision trees.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/multiclass_classifier/random_forest.rst]: A random forest

**Comment by karlnapf** [comment]
I guess this is almost ready, needs updated data version.
@Saurabh7 pls check the travis build and tell me once it executed the integration test for this examples
Then merge

**Comment by karlnapf** [comment]
needs rebase,
any updates?

**Comment by Saurabh7** [comment]
updated, need to check travis.

**Comment by Saurabh7** [comment]
299/424 Test #277: integration_meta_cpp-multiclass_classifier-random_forest .....................   Passed    0.10 sec

**Comment by Saurabh7** [comment]
Actually, I think it might be need to set num_threads explicitly to 1 here too.

Adding `rand_forest.parallel.set_num_threads(1)`  in  random_forest.sg results in errors in the next line.

```
shogun/build/examples/meta/cpp/multiclass_classifier/random_forest.cpp: In function ‘int main(int, char**)’:
shogun/build/examples/meta/cpp/multiclass_classifier/random_forest.cpp:19:1: error: ‘rand_forest’ was not declared in this scope
 rand_forest->train(features_train);
 ^

```

**Comment by karlnapf** [comment]
Nice!
Can we have regression as well?

**Comment by Saurabh7** [comment]
added it here: https://github.com/shogun-toolbox/shogun/pull/3360

**Pull request merged**